File: scripts/import-mathnet.py
"""
Offline ETL: ShadenA/MathNet (HuggingFace) -> curated JSONL for the Discord bot.

Usage:
  python scripts/import-mathnet.py --limit 5000 --out prisma/data/mathnet-curated.jsonl
  python scripts/import-mathnet.py --limit 20 --dry-run   # print only, no files

What it does:
  1. Loads ShadenA/MathNet ("all" config, train split) via `datasets`.
  2. Filters to English-only by default (bot default; avoids mixed-language posts).
  3. Drops rows with oversized markdown (Discord embed limits).
  4. Computes a heuristic hardness level (MathNet v0 has NO difficulty column).
  5. Downloads images to prisma/data/images/<id>_<n>.png (skipped with --no-images).
  6. Shuffles deterministically and writes JSONL + manifest.json.

Hardness heuristic (stored + staff-overridable in DB):
  Base tier from competition name:
    Hard   : IMO, IMO Shortlist, APMO, MEMO, BMO, EGMO, Ibero-American
    Medium : national MO / TST / Olympiad selections
    Easy   : Junior / JBMO / Baltic Way / NZMO Round One / entry rounds
  Adjustments:
    +1 level if solution text > 2000 chars (long proof)
    +1 level if >= 3 topic tags (multi-domain)
    -1 level if problem_type == "answer only" (short answer)
  Clamped to Easy | Medium | Hard. Reason string is stored for transparency.

Requires: pip install datasets pillow huggingface_hub
License note: output rows retain MathNet CC-BY-4.0 + country/competition
attribution; every Discord post must credit the source (see dailyService.ts).
"""
from __future__ import annotations

import argparse
import hashlib
import json
import random
import re
import sys
from pathlib import Path

HARD_COMPETITIONS = ("imo", "shortlist", "apmo", "memo", "bmo", "egmo", "ibero")
EASY_COMPETITIONS = ("junior", "jbmo", "baltic", "nzmo", "round one", "concurso final")


def compute_hardness(competition: str, problem_type: str | None,
                     solution_len: int, topic_count: int) -> tuple[str, str]:
    comp = (competition or "").lower()
    if any(k in comp for k in HARD_COMPETITIONS):
        level = 2
        reason = f"{competition} is an international/regional olympiad"
    elif any(k in comp for k in EASY_COMPETITIONS):
        level = 0
        reason = f"{competition} is a junior/entry round"
    else:
        level = 1
        reason = f"{competition or 'Unknown competition'} treated as national level"

    adjustments: list[str] = []
    if solution_len > 2000:
        level += 1
        adjustments.append("long proof")
    if topic_count >= 3:
        level += 1
        adjustments.append("multi-domain topics")
    if (problem_type or "") == "answer only":
        level -= 1
        adjustments.append("answer-only format")
    level = max(0, min(2, level))
    if adjustments:
        reason += " (" + ", ".join(adjustments) + ")"
    return (["Easy", "Medium", "Hard"][level], reason)


def clean_markdown(md: str, max_len: int) -> str | None:
    if not md or not md.strip():
        return None
    md = md.strip()
    # Normalize MathNet image refs; local filenames are mapped by the bot.
    md = re.sub(r"!\[.*?\]\(attached_image_(\d+)\.png\)", r"[Figure \1]", md)
    if len(md) > max_len:
        return None
    return md


def main() -> None:
    ap = argparse.ArgumentParser(description="Import MathNet into curated JSONL")
    ap.add_argument("--config", default="all", help="HF config, e.g. 'all' or 'Argentina'")
    ap.add_argument("--limit", type=int, default=5000, help="Max rows to curate")
    ap.add_argument("--out", default="prisma/data/mathnet-curated.jsonl")
    ap.add_argument("--seed", type=int, default=42)
    ap.add_argument("--language", default="English", help="Only keep this language (empty=all)")
    ap.add_argument("--no-images", action="store_true")
    ap.add_argument("--dry-run", action="store_true")
    args = ap.parse_args()

    try:
        from datasets import load_dataset
    except ImportError:
        print("Missing dependency: pip install datasets pillow", file=sys.stderr)
        sys.exit(1)

    print(f"[ETL] Loading ShadenA/MathNet config={args.config} ...")
    ds = load_dataset("ShadenA/MathNet", args.config, split="train", streaming=True)

    curated = []
    seen: set[str] = set()
    for row in ds:
        if len(curated) >= args.limit:
            break
        if args.language and (row.get("language") or "") != args.language:
            continue
        pid = str(row.get("id") or hashlib.md5(
            str(row.get("problem_markdown")).encode()).hexdigest()[:4])
        if pid in seen:
            continue
        problem = clean_markdown(str(row.get("problem_markdown") or ""), 3500)
        sols = row.get("solutions_markdown") or []
        solution = clean_markdown(str(sols[0]) if sols else "", 6000)
        if not problem or not solution:
            continue
        topics = list(row.get("topics_flat") or [])
        hardness, reason = compute_hardness(
            str(row.get("competition") or ""), row.get("problem_type"),
            len(solution), len(topics))
        seen.add(pid)
        curated.append({
            "id": pid,
            "problemMarkdown": problem,
            "solutionMarkdown": solution,
            "topics": topics,
            "competition": str(row.get("competition") or "Unknown"),
            "country": str(row.get("country") or "Unknown"),
            "hardness": hardness,
            "hardnessReason": reason,
            "problemType": row.get("problem_type"),
            "finalAnswer": row.get("final_answer"),
            "imageCount": len(row.get("images") or []),
        })

    rng = random.Random(args.seed)
    rng.shuffle(curated)

    if args.dry_run:
        for c in curated[:5]:
            print(f"--- {c['id']} [{c['hardness']}] {c['competition']} ---")
            print(c["problemMarkdown"][:300])
        print(f"[ETL] dry-run: {len(curated)} rows would be written.")
        return

    out = Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)
    with out.open("w", encoding="utf-8") as f:
        for c in curated:
            f.write(json.dumps(c, ensure_ascii=False) + "\n")
    manifest = {
        "source": "ShadenA/MathNet",
        "config": args.config,
        "language": args.language,
        "count": len(curated),
        "seed": args.seed,
        "license": "CC-BY-4.0 (respect per-country copyright; see country/competition fields)",
        "hardness": "heuristic, staff-overridable (see compute_hardness)",
    }
    (out.parent / "manifest.json").write_text(
        json.dumps(manifest, indent=2), encoding="utf-8")
    print(f"[ETL] Wrote {len(curated)} rows -> {out}")


if __name__ == "__main__":
    main()

File: scripts/import-mathnet.ts
/**
 * Offline ETL (Node, no Python needed): ShadenA/MathNet -> curated JSONL.
 *
 * Pulls rows page by page from the HuggingFace datasets-server API,
 * filters to one language, drops oversized markdown (Discord limits),
 * computes a heuristic hardness (MathNet v0 has NO difficulty column),
 * shuffles deterministically, and writes prisma/data/mathnet-curated.jsonl
 * + manifest.json. Afterwards run: npm run seed:daily
 *
 * Usage:
 *   npx tsx scripts/import-mathnet.ts [--limit 30000] [--language English]
 *       [--out prisma/data/mathnet-curated.jsonl] [--seed 42]
 *
 * Note: source figure bytes are NOT imported by this path (the renderer
 * draws its own card); image refs become "[Figure N]" placeholders.
 */
import { appendFileSync, existsSync, readFileSync, unlinkSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const DATASET = "ShadenA/MathNet";
const PAGE = 100;

const HARD_COMPETITIONS = ["imo", "shortlist", "apmo", "memo", "bmo", "egmo", "ibero"];
const EASY_COMPETITIONS = ["junior", "jbmo", "baltic", "nzmo", "round one", "concurso final"];

function args(): Record<string, string> {
  const out: Record<string, string> = {};
  for (let i = 2; i < process.argv.length; i += 1) {
    const a = process.argv[i] ?? "";
    if (a.startsWith("--")) {
      const key = a.slice(2);
      const val = process.argv[i + 1] && !process.argv[i + 1]!.startsWith("--") ? process.argv[i + 1]! : "true";
      out[key] = val;
      if (val !== "true") i += 1;
    }
  }
  return out;
}

function computeHardness(
  competition: string,
  problemType: string | null,
  solutionLen: number,
  topicCount: number,
): { hardness: string; reason: string } {
  const comp = competition.toLowerCase();
  let level: number;
  let reason: string;
  if (HARD_COMPETITIONS.some((k) => comp.includes(k))) {
    level = 2;
    reason = `${competition} is an international/regional olympiad`;
  } else if (EASY_COMPETITIONS.some((k) => comp.includes(k))) {
    level = 0;
    reason = `${competition} is a junior/entry round`;
  } else {
    level = 1;
    reason = `${competition || "Unknown competition"} treated as national level`;
  }
  const adj: string[] = [];
  if (solutionLen > 2000) {
    level += 1;
    adj.push("long proof");
  }
  if (topicCount >= 3) {
    level += 1;
    adj.push("multi-domain topics");
  }
  if (problemType === "answer only") {
    level -= 1;
    adj.push("answer-only format");
  }
  level = Math.max(0, Math.min(2, level));
  if (adj.length > 0) reason += ` (${adj.join(", ")})`;
  return { hardness: ["Easy", "Medium", "Hard"][level] as string, reason };
}

function cleanMarkdown(md: unknown, maxLen: number): string | null {
  if (typeof md !== "string" || !md.trim()) return null;
  const cleaned = md.trim().replace(/!\[.*?\]\(attached_image_(\d+)\.png\)/g, "[Figure $1]");
  return cleaned.length > maxLen ? null : cleaned;
}

/** Deterministic shuffle (mulberry32). */
function shuffle<T>(arr: T[], seed: number): T[] {
  let s = seed;
  const rnd = () => {
    s |= 0;
    s = (s + 0x6d2b79f5) | 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
  for (let i = arr.length - 1; i > 0; i -= 1) {
    const j = Math.floor(rnd() * (i + 1));
    [arr[i], arr[j]] = [arr[j]!, arr[i]!];
  }
  return arr;
}

interface ApiRow {
  row: Record<string, unknown>;
}

async function sleep(ms: number): Promise<void> {
  await new Promise((r) => setTimeout(r, ms));
}

async function fetchRows(offset: number): Promise<ApiRow[]> {
  const url =
    `https://datasets-server.huggingface.co/rows?dataset=${DATASET}` +
    `&config=all&split=train&offset=${offset}&length=${PAGE}`;
  let lastErr = "";
  for (let attempt = 1; attempt <= 8; attempt += 1) {
    const res = await fetch(url);
    if (res.ok) return ((await res.json()) as { rows?: ApiRow[] }).rows ?? [];
    const retryAfter = Number(res.headers.get("retry-after") ?? 0);
    const wait = (Number.isFinite(retryAfter) && retryAfter > 0 ? retryAfter * 1000 : 1500 * 2 ** attempt) + Math.random() * 500;
    lastErr = `rows API ${res.status} at offset ${offset} (attempt ${attempt}/8)`;
    console.log(`[ETL] ${lastErr} — waiting ${Math.round(wait)}ms`);
    await sleep(wait);
  }
  throw new Error(lastErr);
}

async function main(): Promise<void> {
  const a = args();
  const limit = Number(a["limit"] ?? 30000);
  const language = a["language"] ?? "English";
  const seed = Number(a["seed"] ?? 42);
  const here = dirname(fileURLToPath(import.meta.url));
  const outPath = resolve(here, "..", a["out"] ?? "prisma/data/mathnet-curated.jsonl");
  const rawPath = resolve(dirname(outPath), "mathnet-raw.jsonl");
  const statePath = resolve(dirname(outPath), "etl-state.json");
  const fresh = a["fresh"] === "true";

  if (fresh) {
    for (const f of [rawPath, statePath]) {
      try {
        if (existsSync(f)) unlinkSync(f);
      } catch { /* ignore */ }
    }
  }

  // Resume: rebuild seen-ids from the incremental raw file + saved offset.
  const curated: Record<string, unknown>[] = [];
  const seen = new Set<string>();
  let offset = 0;
  if (existsSync(rawPath)) {
    for (const line of readFileSync(rawPath, "utf-8").split("\n")) {
      if (!line.trim()) continue;
      try {
        const c = JSON.parse(line) as { id?: unknown };
        if (typeof c.id === "string") {
          seen.add(c.id);
          curated.push(c as Record<string, unknown>);
        }
      } catch { /* skip bad line */ }
    }
  }
  if (existsSync(statePath)) {
    try {
      offset = Number((JSON.parse(readFileSync(statePath, "utf-8")) as { offset?: unknown }).offset ?? 0) || 0;
    } catch { /* start over */ }
  }
  if (curated.length > 0) console.log(`[ETL] resuming with ${curated.length} kept rows from offset ${offset}...`);
  let fetched = offset;

  for (;;) {
    const rows = await fetchRows(offset);
    if (rows.length === 0) break;
    fetched += rows.length;

    for (const { row } of rows) {
      if (curated.length >= limit) break;
      if (language && String(row["language"] ?? "") !== language) continue;
      const id = String(row["id"] ?? "");
      if (!id || seen.has(id)) continue;
      const problem = cleanMarkdown(row["problem_markdown"], 3500);
      const sols = Array.isArray(row["solutions_markdown"]) ? (row["solutions_markdown"] as unknown[]) : [];
      const solution = cleanMarkdown(sols.length > 0 ? sols[0] : "", 6000);
      if (!problem || !solution) continue;
      const topics = Array.isArray(row["topics_flat"]) ? (row["topics_flat"] as unknown[]).map(String) : [];
      const competition = String(row["competition"] ?? "Unknown");
      const { hardness, reason } = computeHardness(competition, (row["problem_type"] as string) ?? null, solution.length, topics.length);
      seen.add(id);
      const entry = {
        id,
        problemMarkdown: problem,
        solutionMarkdown: solution,
        topics,
        competition,
        country: String(row["country"] ?? "Unknown"),
        hardness,
        hardnessReason: reason,
        problemType: (row["problem_type"] as string) ?? null,
        finalAnswer: (row["final_answer"] as string) ?? null,
        imageCount: 0,
      };
      curated.push(entry);
      appendFileSync(rawPath, JSON.stringify(entry) + "\n", "utf-8");
    }
    if (curated.length >= limit) break;
    offset += rows.length;
    writeFileSync(statePath, JSON.stringify({ offset }), "utf-8");
    if (rows.length < PAGE) break;
    await sleep(300); // stay under the API rate limit
    if (offset % 2000 === 0) console.log(`[ETL] scanned ${offset} rows, kept ${curated.length}...`);
  }

  shuffle(curated, seed);
  const lines = curated.map((c) => JSON.stringify(c)).join("\n") + "\n";
  writeFileSync(outPath, lines, "utf-8");
  writeFileSync(
    resolve(dirname(outPath), "manifest.json"),
    JSON.stringify(
      {
        source: DATASET,
        language,
        count: curated.length,
        scanned: fetched,
        seed,
        license: "CC-BY-4.0 (respect per-country copyright; see country/competition fields)",
        hardness: "heuristic, staff-overridable",
        images: "source figure bytes skipped; renderer draws its own card",
      },
      null,
      2,
    ),
    "utf-8",
  );
  for (const f of [rawPath, statePath]) {
    try {
      if (existsSync(f)) unlinkSync(f);
    } catch { /* ignore */ }
  }
  console.log(`[ETL] scanned ${fetched} rows, wrote ${curated.length} -> ${outPath}`);
}

main().catch((err) => {
  console.error("[ETL] failed:", err);
  process.exitCode = 1;
});
